[cpp.line] Address string-literal vs s-char-sequence

AlisdairM opened the pull request:
The text of this specification places a restriction on a _string-literal_ that is not present in the grammar, although the grammar is an exact match for a _string-literal_ with that restriction.

The restricted term is also used when specifying the effects of the line directive with a string, so prefer to move _string-literal_ into the grammar rather than strike that paragraph and adjust the specification for its effects.

Added a cross-reference to where character string literal is defined as it may not be obvious to all readers that this is a term of power, even though it is defined as a term and has an index entry.

AlisdairM committed:
[cpp.line] Address string-literal vs s-char-sequence

The text of this specification places a restriction on a string-literal
that is not present in the grammar, although the grammar is an exact
match for a string-literal with that restriction.

The restricted term is also used when specifying the effects of the
line directive with a string, so prefer to move string-literal into
the grammar rather than strike that paragraph and adjust the effects.

Added a cross-reference to where character string literal is defined
as it may not be obvious to all readers that this is a term of power,
even though it is defined as a term and has an index entry.

jensmaurer commented on source/preprocessor.tex:
```diff
@@ -2102,7 +2102,7 @@
 A preprocessing directive of the form
 \begin{ncsimplebnf}
-\terminal{\# line} digit-sequence \terminal{"} \opt{s-char-sequence} \terminal{"} new-line
+\terminal{\# line} digit-sequence \terminal{"} \opt{string-literal} \terminal{"} new-line
```
No.

First, a _string-literal_ includes the delimiting quotation marks, so this is a bad change.
Second, a _string-literal_ allows for an _encoding-prefix_ or rawness, but the status quo "form" match does not.  So, this is a bad change, because it would appear to allow such prefixes (and we don't know what they mean in this context).  A reference to "the character string literal" does not establish a suitable constraint.

Yes, p1 talks about a _string-literal_ with zero context, and that's a bug that needs to be fixed, but this intro sentence is at issue, not the rest of the specification without further elaboration.  What does C23 say in that area?

AlisdairM committed:
[lex.string] Adopt plain-string-literal grammar to simplify string literals

There are several places in the grammar where we want to refer to a
plain character string literal without a prefix and that is not raw.
Introduce _character-string_ into the grammar to replace a quoted
_s-char-sequence_ in a number of places, and move the `R` of a raw
string literal into the _raw-string_ grammar so that _raw-string_
and _character-string_ are used consistently where they appear.
